// slack/index.js
const request = require('request');

module.exports = function(options) {

  this.user = 'default';

  this.channel = '';

  this.request = function(path, query, next) {
    query.token = 'xoxp-89670216483-89737576834-89670820275-f424ea0798674acb6f0b33f96f1760d5';
    var queryArray = [];
    for(q in query) { queryArray.push(`${q}=${query[q]}`); }
    var url = `https://slack.com/api${path[0] == '/' ? '' : '/'}${path}?${queryArray.join('&')}`;
    request(url, function(err, response, body) {
      body = JSON.parse(body);
      if(err) return next(err);
      if(!body.ok) return next(body.error);
      next(err, body);
    })
  }

  this.postMessage = function(text, next){
    this.request('chat.postMessage', { username: this.user, channel: this.channel, text: text }, next);
  }

  this.listChannels = function(next){
    this.request('channels.list', { exclude_archived: 1 }, next);
  }

  this.listUsers = function(next){
    this.request('users.list', {}, next);
  }

  this.searchMessages = function(query, next) {
    this.request('search.messages', { query: query }, next);
  }

  return this;
};
